Automatically close timepicker upon selection

lukasolson opened the pull request:
This PR automatically closes the timepicker when you select a time window.

The first stab at this just added a call to `kbnTopNav.toggle('filter')` at the end of the `setQuick` method, but this was problematic because the timepicker would disappear before the `$scope` variable changes propagated to the actual time filter. Wrapping the `toggle` call in a `$timeout` seemed to solve the problem, but I didn't like this solution.

Instead, I decided to change the way we reference the `from` and `to` values in the timepicker. Instead of having a local `from` and `to` in the timepicker's `$scope`, I decided to directly reference the time filter's `time` (which has its own `from` and `to`) and directly modify them. This way, I don't have to wrap the call to `toggle` in a `$timeout` and the changes still propagate. 

Closes https://github.com/elastic/kibana/issues/9188.

tbragin commented:
Checked it out and it looks good. Amazing how such a simple change can improve the experience. 

After looking at the timepicker closer, I realized we have "Go" buttons in the Relative and Absolute selectors. What do you think about auto-closing when user clicks those buttons as well?

weltenwort commented on src/ui/public/timepicker/timepicker.js:
```diff
@@ -20,14 +20,14 @@ module.directive('kbnTimepicker', function (quickRanges, timeUnits, refreshInter
     },
     template: html,
-    controller: function ($scope) {
+    require: '^kbnTopNav',
```
I would advise to avoid introducing such a dependency on the hierarchy the timepicker is placed in, because it limits reuse and couples the components more tightly. Why not add a `hide: '&'` attribute that can be used by the parent component to pass a function as in `<kbn-timepicker ... hide="kbnTopNav.close('filter')">`. This way the `<kbn-timepicker>` directive can be used in contexts where there is no `kbnTopNav`.

Even better would be to move from implicitly modifying the values to an explicit callback (e.g. `change(newFrom, newTo)` with `change` having been passed in by the parent) and let the parent control the visibility, but that might be out of scope for this PR.

Bargs commented on src/ui/public/chrome/config/filter.html:
```diff
@@ -1,6 +1,5 @@
-  from="timefilter.time.from"
-  to="timefilter.time.to"
+  time="timefilter.time"
   mode="timefilter.time.mode"
```
Any reason you removed `from` and `to` but not `mode`?

lukasolson commented:
I've updated this PR with the suggested updates:

The time picker no longer requires the `kbnTopNav` in a parent scope (thanks, @weltenwort).
Instead of being passed in `timefilter.time` and `timefilter.time.mode`, the `mode` now is read from `timefilter.time` instead (thanks, @Bargs).
The time picker is now automatically closed when you apply a relative or absolute time range as well (thanks, @tbragin).

Could you guys take another look?

tbragin commented:
For some reason, when I checked out the latest PR against master, I got `server    log   [01:19:11.910] [error][status][plugin:elasticsearch@5.2.0] Status changed from yellow to red - This version of Kibana requires Elasticsearch v5.2.0 on all nodes. I found the following incompatible nodes in your cluster: v6.0.0-alpha1 @ 127.0.0.1:9200 (127.0.0.1)` on startup.

weltenwort commented:
@tbragin, sounds like the `version` in the `package.json` of your working copy has been set to `5.2.0` - maybe contamination from some previous checkout?

weltenwort commented on src/ui/public/chrome/config/filter.html:
```diff
@@ -1,6 +1,6 @@
   active-tab="'filter'"
-  interval="timefilter.refreshInterval">
+  interval="timefilter.refreshInterval"
+  on-select="kbnTopNav.toggle('filter')">
```
Is there any situation in which we really want to "toggle"? I feel like the intent would be better expressed by using `kbnTopNav.close('filter')`.

lukasolson replied:
Great point. I've updated the PR.

Bargs commented:
Aside from the CI test failures, this is looking good to me.

lukasolson commented:
Note: Something I neglected to test that I actually broke with this PR was when you selected the "interval" tab of the timepicker. I've fixed the test @Bargs mentioned and fixed this bug. If you guys could take one more quick look that'd be great 😊

Bargs commented:
Hmmm jenkins still seems to be unhappy

weltenwort commented:
Seems to be an unrelated timing issue in the test? The code LGTM.

lukasolson commented:
As I was playing with this I realized, when users have the "Auto-refresh" tab selected and select a refresh interval, it would probably make sense to close the time picker, right?

@Bargs @weltenwort @tbragin

lukasolson commented:
@Bargs I've finally got those pesky tests fixed. Could you take one more look?

Bargs commented:
@lukasolson @spalger it looks like there's a linter error after the master merge

lukasolson commented:
@Bargs Fixed.

tbragin commented:
@lukasolson good catch on auto-refresh. yes, i think it makes sense to close it after a selection is made in that screen as well.

tbragin commented:
@cjcenizal I copied you on the original issue, but didn't see a reaction there. hopefully all is good with you when it comes to this change?

cjcenizal commented:
How sure are we that this solves a problem for users without introducing a new problem? Is it possible users would want to keep the time picker open, e.g. to select different time ranges and observe the change in data? (I hate to be a stickler because I personally like this change, but I have to ask. :)

tbragin commented:
Hey @cjcenizal - That's a fair question to ask, and that's exactly why I CC'd you! 

We did discuss user's intentions when we considered this change. In majority of cases, we believe when a user makes a selection in time picker and selects a time range they intended, they would like for the time picker to close at that point. 

However there are situations, where it's advantageous to keep the time picker open, so let's talk about some of them. 
* The first situation I can think of is when the user wants to keep changing the time. An example of that might be navigating forward and backward in time. Luckily, for that purpose we just introduced [previous / next time navigation buttons](https://github.com/elastic/kibana/pull/9253) in time picker that are a better solution to that problem than manually changing the time frame repeatedly. 
* The other main use case i can imagine is when a user makes a selection that is too wide or too narrow. In the case of too wide, I think the user can simply brush one of date histogram based charts on the screen to the time range they want, assuming that is present (it is in Discover and in most dashboards based on time-based data). Zooming out is harder (at least until/if we introduce [zoom out controls](https://github.com/elastic/kibana/issues/9434)), but for now, say if you selected "last 2 years" and see no data, and realize that your data is actually 3 years back -- you can still use the previous button to get to where you need to be with one click. 

In summary, we do think that in majority of use cases, this change will be beneficial, and the saved clicks in cases where user made the proper selection the first time will outweigh the few added clicks when the user made a mistake. 

Does that make sense?

cjcenizal commented:
@tbragin Thanks for the analysis! Sounds to me like this is a good change to make.

cjcenizal commented:
Does this address https://github.com/elastic/kibana/issues/5682 ?

lukasolson commented:
@cjcenizal I believe so, although I don't remember that ever being the expected behavior before this PR... Strange.

lukasolson commented:
Replaced by https://github.com/elastic/kibana/pull/9618 (less invasive solution).
